### DISCUSSION TITLE
Building SuiteSparse with Intel C++ compiler

### DISCUSSION THREAD

**Discussion by ChessMastery** [DrTimothyAldenDavis/SuiteSparse#869] [Q&A] [answered]
I want to build SuiteSparse on Windows 10 with Intel C++ compiler instead of MSVC (as I've got an intel processor and MSVC appears to be bad in vectorization). Previously I built it with MSVC except SPEX (I don't need it).
I try to specify the compiler by typing in Intel oneAPI command prompt
**cmake -DCMAKE_C_COMPILER=icx -DCMAKE_CXX_COMPILER==icx ..**
or
**cmake -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER==g++ ..**
However, the output is still
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.36.32535.0
-- The CXX compiler identification is MSVC 19.36.32535.0
icx, gcc and g++ commands works, I suppose that means the compilers are installed in the system. What should I do in order to build SuiteSparse with Intel compiler? Intel C++ compiler can be set in Visual Studio GUI, so MSVC in not the only C++ compiler supported by Visual Studio 17 2022 build system.

**Comment by mmuetzel** [comment, marked as answer]
Are you using a clean build tree? If you built with MSVC in that folder before, change to a different folder or delete its content.
If I'm reading my notes correctly, the following commands worked for me in the past:
```
cmake -G"Ninja Multi-Config" -DCMAKE_CXX_COMPILER=icx-cl -DCMAKE_C_COMPILER=icx-cl -DCMAKE_Fortran_COMPILER=ifx -DSUITESPARSE_ENABLE_PROJECTS="suitesparse_config;mongoose;amd;btf;camd;ccolamd;colamd;cholmod;cxsparse;ldl;klu;umfpack;paru;rbio;spqr" -DSUITESPARSE_USE_CUDA=OFF --fresh ..
cmake --build . --config Release
```

**Reply by ChessMastery** [reply]
I use a clean build tree. Thanks, with Ninja generator it works for me as well. But I wonder whether one can specify Intel C++ compiler for Visual Studio build system, as I have problems building other projects (not SuiteSparse) with such an issue.

**Reply by mmuetzel** [reply]
I haven't tested that. But have you tried using `icx-cl` instead of `icx` with the Visual Studio generator? I'd guess that that generator might need a compiler that is command line compatible to `cl`.

Just out of interest: Why would you need to build with the Visual Studio generator if it works correctly with the "Ninja Multi-Config" generator?

**Reply by ChessMastery** [reply]
No, it did not work, MSBuild still uses MSVC when I specify icx-cl compilers. Well, I am new to build systems like cmake and I have not yet figured out how exactly do they work, but I guess when you build high-performance library, there might be differencies when using different generators (for example, one generator may not support some compilers or compiler versions). So why not build it with MSBuild and Ninja and benchmark? Moreover, I guess it might be that the cmake scripts are written for MSBuild specifically, and Ninja cannot be used or would be inproperly configured if used (or it would take a lot of time to modify cmake scripts for Ninja). Please correct me if I am wrong.

**Reply by mmuetzel** [reply]
Reading parts if the documentation of that generator by CMake and by Visual Studio, I'm not sure if it can use different compilers than the MS C compilers at all. It also seems to only work with Fortran compilers from Intel.

I've never heard of any difference in performance of the produced binaries when using different generators. I'd be inclined to think that it would be considered as a bug in CMake if there were any such differences.
The performance of the build process itself can obviously depend on the generator though.

If you are worried about potential issues with different CMake generators, you could run your comparison using the MS compiler (and build once with the Visual Studio generator and then again with the Ninja generator).
